### ISSUE TITLE
Extend DataLoader with transform arguments to keep that logic in the DataModule

### ISSUE THREAD

**Issue by NumesSanguis** [PyTorchLightning/pytorch-lightning#3148]
## 🚀 Feature

Extend DataLoader to accept transform.
Possibly split into 2 arguments:
1. sample_transform
2. batch_transform

The new process of data retrieval would go as follow:
1. Retrieve sample tuple with (input, target)
2. sample_transform(sample)
3. Concatenate to batch
4. batch_transform(batch)

### Motivation

The new DataModule [page of the documentation](https://pytorch-lightning.readthedocs.io/en/stable/datamodules.html#train-dataloader) incorrectly includes a `transform=transforms` argument for the `DataLoader`:

```python
class MNISTDataModule(pl.LightningDataModule):
    def train_dataloader(self):
        transforms = transform_lib.Compose([
            transform_lib.ToTensor(),
            transform_lib.Normalize(mean=(0.5,), std=(0.5,)),
        ])
        return DataLoader(self.train_dataset, transform=transforms, batch_size=64)
```

This extra argument would allow to truly separate the `Dataset` and `DataModule` in terms of functionality.

If we do transforms on the samples after retrieving it from the Dataset, we could also logically include not only sample transforms, but also possible batch transforms.

### Pitch

`Dataset` and `DataModule` are not truly separated in functionality yet.
While the documentation suggests that transforms should be specified in `DataModule`, in the end the transform is still passed on to `Dataset`.
This requires `Dataset` to be transform-aware. If a `Dataset` is not configured with something similar to:
```python
if self.transform:
    input = self.transform(input)
```
, it will not be possible to transform the data.
If it is necessary to write transform code in the `Dataset` class, why not specify it directly there?

Extra 1: If `Dataset` does not have to be tuned to expect PyTorch transforms, this Dataset could be used for both PyTorch and scikit-learn without encoding specific logic for this in the `Dataset` class.

Extra 2: If we move transform to DataLoader, we can also both specify per-sample transforms and per-batch transforms.

p.s. Would it be faster to transform a batch of numpy arrays into Tensor, or is there no different with transforming the arrays 1-by-1 into a Tensor first before batching them?

### Alternatives

Keep it as-is, passing the transform to a `Dataset` and require the user to make sure the `Dataset` can handle potential transforms.

### Additional context

Liked by: @Borda 
Others mentioned by him: @nateraw @awaelchli @williamFalcon

**Comment by Borda** [comment]
I like your point to decouple also this part of the pipeline as most DL training contains some augmentations...
cc: @PyTorchLightning/core-contributors

**Comment by NumesSanguis** [comment]
This split in functionality can already be found in the PyTorch based FastAI(2) library in their [DataBlock](https://docs.fast.ai/data.block.html#DataBlock). This class accepts a `item_tfms` and a `batch_tfms` argument.

FastAI also has support for something they call [presizing](https://github.com/fastai/fastbook/blob/master/05_pet_breeds.ipynb):

> ...
> The challenge is that, if performed after resizing down to the augmented size, various common data augmentation transforms might introduce spurious empty zones, degrade data, or both. For instance, rotating an image by 45 degrees fills corner regions of the new bounds with emptyness, which will not teach the model anything. Many rotation and zooming operations will require interpolating to create pixels. These interpolated pixels are derived from the original image data but are still of lower quality.
>
> To work around these challenges, presizing adopts two strategies ...:
>
>  1.  Resize images to relatively "large" dimensions—that is, dimensions significantly larger than the target training dimensions.
>  2.  Compose all of the common augmentation operations (including a resize to the final target size) into one, and perform the combined operation on the GPU only once at the end of processing, rather than performing the operations individually and interpolating multiple times.
>
> The first step, the resize, creates images large enough that they have spare margin to allow further augmentation transforms on their inner regions without creating empty zones. This transformation works by resizing to a square, using a large crop size. On the training set, the crop area is chosen randomly, and the size of the crop is selected to cover the entire width or height of the image, whichever is smaller.
>
> In the second step, the GPU is used for all data augmentation, and all of the potentially destructive operations are done together, with a single interpolation at the end.
>
> ![](https://raw.githubusercontent.com/fastai/fastbook/efff9626b09eb5ae3d83449c3830856abe6d8d53/images/att_00060.png)
> 
> This picture shows the two steps:
>
> 1.  Crop full width or height: This is in item_tfms, so it's applied to each individual image before it is copied to the GPU. It's used to ensure all images are the same size. On the training set, the crop area is chosen randomly. On the validation set, the center square of the image is always chosen.
> 2.  Random crop and augment: This is in batch_tfms, so it's applied to a batch all at once on the GPU, which means it's fast. On the validation set, only the resize to the final size needed for the model is done here. On the training set, the random crop and any other augmentations are done first.

By having separated the transforms from `Dataset`, it might become possible for PyTorch Lightning to speed up transforms by smartly, under the hood, combining all transforms into 1, as done by FastAI.

**Comment by williamFalcon** [comment]
So, we don’t enforce transforms to be passed in but they CAN be passed in via init. Why wouldn’t this work?

init(self, train_transform, etc...)  

the datamodule is just an interface. the ones we codes have default transforms. and in fact, you can set them however you want or pass them in. 

dm = DataModule()
dm.train_transforms = ...

or

dm = DataModule(train_transform)

**Comment by justusschock** [comment]
@williamFalcon because there is no defined way, where these transforms will be called. In most cases they are transforms for the dataset. But here the request is to have them as part of the data loader

**Comment by williamFalcon** [comment]
got it. 

ok, unlikely that we will add a custom dataloader to lightning since the api is finalized and needs to stay thin. 

but we can consider for another project

**Comment by NumesSanguis** [comment]
The issue is not that it cannot be passed to the DataModule. The issue is a conceptual one. The `Dataset` class should be disconnected from possible transforms, but now the `DataModule` still has to pass it on to the `Dataset` which also has to know how to use that transform, before the transform can be used.
How to transform data should not be handled in the `Dataset` class.

Although I can see this is at a deeper level that should be tackled from PyTorch, and not Lightning. The `DataModule` would be a perfect fit for this separation however.

**Comment by JVGD** [comment]
I very much agree with this Issue philosophy. Conceptually speaking, it is very clean to have the Dataset logic focus on loading the data and the DataModule to focus on augment & normalize it. When we started porting our code to Pytorch Lightning, about 2 weeks ago, we considered this approach in or designs:
![image](https://user-images.githubusercontent.com/6736158/95191129-99365980-07d0-11eb-97cf-76b58c136d6c.png)

With this approach, the data folks can load or import the data from wherever they want and we, from the AI team, can define how that data will be augmented, transformed & normalized. So basically is just a matter of interfacing between 2 teams and between 2 logically separated tasks.

We are sticking to our designs because it is almost a must for us but it would be really nice to have this the standard library

**Comment by williamFalcon** [comment]
in bolts we separated the transforms from the data:

https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/pl_bolts/models/self_supervised/simclr/simclr_module.py#L248

**Comment by dboeckenhoff** [comment]
> in bolts we separated the transforms from the data:
> 
> https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/pl_bolts/models/self_supervised/simclr/simclr_module.py#L248

Now @ https://github.com/Lightning-Universe/lightning-bolts/blob/master/src/pl_bolts/models/self_supervised/simclr/simclr_module.py
